doc/lock_files.md

```diff
@@ -0,0 +1,239 @@
+* Compiler override
+* `extra-deps`
+* Flags
+* GHC options
+* Hidden packages
+
+The resolver can either specify a compiler version or another snapshot
+file. This snapshot file can contain the same information referenced
+above for a `stack.yaml`, with the following differences:
+
+* The `extra-deps` are called `packages`
+* Drop packages can be included
+
+Some of this information is, by its nature, complete. For example, the
+"flags" field cannot be influenced by anything outside of the file
+itself.
+
+On the other hand, some information in these files can be
+incomplete. Consider:
+
+```yaml
+resolver: lts-13.9
+packages: []
+extra-deps:
+- https://hackage.haskell.org/package/acme-missiles-0.3.tar.gz
+```
+
+This information is _incomplete_, since the contents of that URL may
+change in the future. Instead, you could specify enough information in
+the `stack.yaml` file to fully resolve that package. That looks like:
+
+```yaml
+extra-deps:
+- size: 1442
+  url: https://hackage.haskell.org/package/acme-missiles-0.3.tar.gz
+  cabal-file:
+    size: 613
+    sha256: 2ba66a092a32593880a87fb00f3213762d7bca65a687d45965778deb8694c5d1
+  name: acme-missiles
+  version: '0.3'
+  sha256: e563d8b524017a06b32768c4db8eff1f822f3fb22a90320b7e414402647b735b
+  pantry-tree:
+    size: 226
+    sha256: 614bc0cca76937507ea0a5ccc17a504c997ce458d7f2f9e43b15a10c8eaeb033
+```
+
+Users don't particularly feel like writing all of that. Therefore,
+it's common to see _incomplete_ information in a `stack.yaml` file.
+
+Additionally, the `lts-13.9` information is _also_ incomplete. While
+we assume in general that LTS snapshots never change, there's nothing
+that technically prohibits that from happening. Instead, the complete
+version of that field is:
+
+```yaml
+resolver:
+  size: 496662
+  url: https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/13/9.yaml
+  sha256: 83de9017d911cf7795f19353dba4d04bd24cd40622b7567ff61fc3f7223aa3ea
+```
+
+Also something people don't feel like writing by hand.
+
+## Recursive snapshot layers
+
+Snapshot files can be _recursive_, where `stack.yaml` refers to
+`foo.yaml`, which refers to `bar.yaml`, which refers to `baz.yaml`. A
+local snapshot file can refer to a remote snapshot file (available via
+an HTTP(S) URL).
+
+We need to encode information from _all_ of these snapshot layers and
+the `stack.yaml` file in the lock file, to ensure that we can detect
+if anything changes.
+
+## Performance
+
+In addition to acting as a pure correctness mechanism, the design of a
+lock file given here also works as a performance improvement. Instead
+of requiring that all snapshot files be fully parsed on each Stack
+invocation, we can store information in the lock file and bypass
+parsing of the additional files in the common case of no changes.
+
+## Lock file contents
+
+The lock file contains the following information:
+
+* The full snapshot definition information, including completed
+  package locations
```

Comment:
Can you clarify here about which "complete package locations" you mean ? Do you mean 
* completed package locations of the packages in the snapshot file ?
* Or complete package locations of the packages in the `extra-deps` file (or both) ?

Reply:
I mean both.